src/Microsoft.ML.Data/DataLoadSave/PartitionedFileLoader.cs

```diff
@@ -0,0 +1,726 @@
+
+                if (Type.HasValue)
+                {
+                    sb.Append($":{Type}");
+                }
+
+                sb.Append($":{Source}");
+
+                return true;
+            }
+        }
+
+        private readonly IHost _host;
+        private readonly IMultiStreamSource _files;
+        private readonly Column[] _columns;
+
+        // Number of tailing directories to include.
+        private readonly int _tailingDirCount;
+
+        // An underlying loader used on each individual loader.
+        private readonly SubComponent<IDataLoader, SignatureDataLoader> _subLoader;
+
+        private readonly IPartitionedPathParser _pathParser;
+
+        private const string RegistrationName = LoadName;
+
+        public PartitionedFileLoader(IHostEnvironment env, Arguments args, IMultiStreamSource files)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            _host = env.Register(RegistrationName);
+            _host.CheckValue(args, nameof(args));
+            _host.CheckValue(files, nameof(files));
+
+            _pathParser = args.PathParserFactory.CreateComponent(_host);
+            _host.CheckValue(_pathParser, nameof(_pathParser), "Factory failed to create a FilePathSpec");
+
+            _subLoader = args.Loader;
+            _files = files;
+
+            string relativePath = GetRelativePath(args.BasePath, files);
+            _columns = ParseColumns(relativePath).ToArray();
+            _tailingDirCount = GetDirectoryCount(relativePath);
+
+            if (args.IncludePathColumn)
+            {
+                var pathCol = new Column()
+                {
+                    Name = "Path",
+                    Source = -1,
+                    Type = DataKind.Text
+                };
+
+                _columns = _columns.Concat(new[] { pathCol }).ToArray();
+            }
+
+            Schema = CreateSchema(_host, _columns, _subLoader);
+        }
+
+        private PartitionedFileLoader(IHost host, ModelLoadContext ctx, IMultiStreamSource files)
+        {
+            Contracts.AssertValue(host);
+            _host = host;
+            _host.AssertValue(ctx);
+            _host.AssertValue(files);
+
+            // ** Binary format **
+            // int: tailing directory count
+            // int: number of columns
+            // foreach column:
+            //   string: column representation
+            // string: subloader
+            // model: file path spec
+
+            _tailingDirCount = ctx.Reader.ReadInt32();
+
+            int numColumns = ctx.Reader.ReadInt32();
+            _host.CheckDecode(numColumns >= 0);
+
+            _columns = new Column[numColumns];
+            for (int i = 0; i < numColumns; i++)
+            {
+                var column = Column.Parse(ctx.LoadString());
+                _host.CheckDecode(column != null);
+                _columns[i] = column;
+            }
+
+            var loader = SubComponent.Parse(ctx.LoadString());
```

Comment:
You're attempting to serialize and deserialize a schema I see. I might prefer that we re-use the old solution to this problem rather than invent what amounts to a new serialization framework.

See for example, the code in `OptionalColumnTransforms` [here](https://github.com/dotnet/machinelearning/blob/c023727b76970ab913ec1ce38276508835c17bcf/src/Microsoft.ML.Transforms/OptionalColumnTransform.cs#L98) or `BinaryLoader` [here](https://github.com/dotnet/machinelearning/blob/c023727b76970ab913ec1ce38276508835c17bcf/src/Microsoft.ML.Data/DataLoadSave/Binary/BinaryLoader.cs#L919). As we see we're reading/writing an empty-dataview with a schema, using the `BinaryLoader`/`Saver`.

Reply:
Although the Schema is built from the Column data it's not limited just to the Schema (maybe this is a shortcoming of how are coded this class). For now, I'm not going to change but will consider refactoring in future iterations to utilize the reusable code.

Reply:
In other scenarios just punting till later may be acceptable, bit in this case we're talking about the loader's model format. Model formats are things we like to have backwards compatible. Unlike other artifacts of the code where change may be acceptable, something here, once checked in, is something we have to live with forever. For this reason I'd appreciate you using an alternate scheme, that does not rely on serializing the command line parsing language. Thanks much!

---
In reply to: [187675779](https://github.com/dotnet/machinelearning/pull/61#discussion_r187675779) [](ancestors = 187675779)

Reply:
Point taken. Investigating further.

Reply:
Resolved